### ISSUE TITLE
Nested ROW projection emulation doesn't work in scalar subqueries

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12930]
I'm running a query like this:

```java
Result<Record2<Integer, Record2<Integer, Integer>>> result = ctx
    .select(
        TAuthor_ID(),
        field(select(row(TBook_ID(), TBook_AUTHOR_ID()))
            .from(TBook())
            .where(TBook_ID().eq(TAuthor_ID().minus(1)))))
    .from(TAuthor())
    .orderBy(TAuthor_ID())
    .fetch();
```

As an integration test for a situation where a nested `ROW` projection is placed in a scalar subquery, expecting it to return `NULL` (see https://github.com/jOOQ/jOOQ/issues/12929). The related issue talks about PostgreSQL, where this works correctly:

```sql
select
  "public"."t_author"."id",
  (
    select row ("public"."t_book"."id", "public"."t_book"."author_id")
    from "public"."t_book"
    where "public"."t_book"."id" = ("public"."t_author"."id" - 1)
  )
from "public"."t_author"
order by "public"."t_author"."id"
```

But in H2 and all other dialects where we emulate this, it doesn't work, producing invalid SQL:

```sql
select
  "PUBLIC"."T_AUTHOR"."ID",
  (
    select "PUBLIC"."T_BOOK"."ID" "row.ID",
    "PUBLIC"."T_BOOK"."AUTHOR_ID" "row.AUTHOR_ID"
    from "PUBLIC"."T_BOOK"
    where "PUBLIC"."T_BOOK"."ID" = ("PUBLIC"."T_AUTHOR"."ID" - 1)
  )
from "PUBLIC"."T_AUTHOR"
order by "PUBLIC"."T_AUTHOR"."ID"
```

Dialects that support `MULTISET` emulations via SQL/JSON or SQL/XML could easily emulate this as well. In other dialects, I don't see how it could be done, currently.

### Rule set

Nested row projections have to be emulated using the `MULTISET` emulation if:

- [x] `MULTISET` emulations are available
- [x] The degree of the row is > 1
- [x] The row expression is projected from a subquery
  - [x] But not from within a row expression predicate, such as `(a, b) IN (SELECT x, y)`. See https://github.com/jOOQ/jOOQ/issues/13129
  - [x] And not from within a derived table. See https://github.com/jOOQ/jOOQ/issues/3497

**Comment by lukaseder** [comment]
The fix works well for actual projections of nested `ROW` values, but it's not trivial to prevent an application of the fix when the consumer of the "projection" is a simple row predicate, such as:

```sql
where t.embeddable = (select u.embeddable from t)
```

In this case, we shouldn't (yet) apply any `MULTISET` related emulation, although it seems possible to add a configuration option to always emulate embeddables using `MULTISET` instead of what we're doing now.

**Comment by lukaseder** [comment]
This isn't really so different from the problems listed here: https://github.com/jOOQ/jOOQ/issues/11824.

Instead of emulating nested rows using `MULTISET` mechanisms only in subqueries, let's switch to the default `MULTISET` emulation everywhere, where `MULTISET` is supported, offering to revert to the "old" emulation with the known issues.

**Comment by lukaseder** [comment]
The case where the embeddable record is projected from within a subquery contained in a predicate (e.g. `t.emb IN (SELECT u.emb FROM u)` would best be identified by implementing https://github.com/jOOQ/jOOQ/issues/3497, where the `Context` can clearly identify the type of subquery we're in.

**Comment by lukaseder** [comment]
Instead of implementing all of #3497, perhaps implementing https://github.com/jOOQ/jOOQ/issues/13129 (`Context::predicandSubquery`) might suffice?

**Comment by lukaseder** [comment]
It looks like we need most of #3497 to avoid breaking embeddable functionality, including whether a subquery is a derived table or a set operation subquery.

**Comment by lukaseder** [comment]
This has been fixed for jOOQ 3.17.0. A backport won't be possible as the fix is very delicate, and relies on a lot of new jOOQ 3.17.0 stuff.
